File: NavigationSampleSafeJava/app/src/main/java/com/websarva/wings/android/navigationsamplesafejava/MemoListFragment.java
package com.websarva.wings.android.navigationsamplesafejava;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.websarva.wings.android.navigationsamplesafejava.data.model.Memo;
import com.websarva.wings.android.navigationsamplesafejava.databinding.FragmentMemoListBinding;
import com.websarva.wings.android.navigationsamplesafejava.databinding.MemoListItemBinding;
import com.websarva.wings.android.navigationsamplesafejava.ui.MemoListViewModel;

import java.util.List;

public class MemoListFragment extends Fragment {
	private FragmentMemoListBinding _fragmentMemoListBinding;
	private MemoListViewModel _memoListViewModel;
	private MemoListAdapter _adapter;

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		ViewModelProvider viewModelProvider = new ViewModelProvider(MemoListFragment.this);
		_memoListViewModel = viewModelProvider.get(MemoListViewModel.class);
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		_fragmentMemoListBinding = FragmentMemoListBinding.inflate(inflater);
		return _fragmentMemoListBinding.getRoot();
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		_adapter = new MemoListAdapter();
		LinearLayoutManager layoutManager = new LinearLayoutManager(requireActivity());
		_fragmentMemoListBinding.rvMenu.setLayoutManager(layoutManager);
		DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(requireActivity(), layoutManager.getOrientation());
		_fragmentMemoListBinding.rvMenu.addItemDecoration(dividerItemDecoration);
		_fragmentMemoListBinding.rvMenu.setAdapter(_adapter);

		LiveData<List<Memo>> _memoListLiveData = _memoListViewModel.getMemoListLiveData();
		MemoListObserver memoListObserver = new MemoListObserver();
		_memoListLiveData.observe(getViewLifecycleOwner(), memoListObserver);
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		_fragmentMemoListBinding = null;
	}

	private class MemoListObserver implements Observer<List<Memo>> {
		@Override
		public void onChanged(List<Memo> memoList) {
			_adapter.submitList(memoList);
		}
	}

	private class MemoListAdapter extends ListAdapter<Memo, MemoViewHolder> {
		public MemoListAdapter() {
			super(new MemoDiff());
		}

		@NonNull
		@Override
		public MemoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			LayoutInflater inflater = LayoutInflater.from(requireActivity());
			MemoListItemBinding binding = MemoListItemBinding.inflate(inflater, parent, false);
			return new MemoViewHolder(binding);
		}

		@Override
		public void onBindViewHolder(@NonNull MemoViewHolder holder, int position) {
			Memo memo = getItem(position);
			holder.bind(memo);
			holder.itemView.setTag(memo.getId());
			holder.itemView.setOnClickListener(new ListItemClickListener());
		}
	}

	private class MemoDiff extends DiffUtil.ItemCallback<Memo> {
		@Override
		public boolean areItemsTheSame(@NonNull Memo oldItem, @NonNull Memo newItem) {
			return oldItem.getId().equals(newItem.getId());
		}

		@Override
		public boolean areContentsTheSame(@NonNull Memo oldItem, @NonNull Memo newItem) {
			return oldItem.getTitle().equals(newItem.getTitle()) || oldItem.getContent().equals(newItem.getContent());
		}
	}

	private class MemoViewHolder extends RecyclerView.ViewHolder {
		private final MemoListItemBinding _binding;

		public MemoViewHolder(MemoListItemBinding binding) {
			super(binding.getRoot());
			_binding = binding;
		}

		public void bind(Memo memo) {
			_binding.memoTitleRow.setText(memo.getTitle());
		}
	}

	private class ListItemClickListener implements View.OnClickListener {
		@Override
		public void onClick(View view) {
			Long memoId = (Long) view.getTag();
			MemoListFragmentDirections.ActionMemoListFragmentToMemoDetailFragment action = MemoListFragmentDirections.actionMemoListFragmentToMemoDetailFragment();
			action.setMemoId(memoId);
			NavController navController = NavHostFragment.findNavController(MemoListFragment.this);
			navController.navigate(action);
		}
	}
}

File: NavigationSampleDSL/app/src/main/java/com/websarva/wings/android/navigationsampledsl/MemoListFragment.kt
package com.websarva.wings.android.navigationsampledsl

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.websarva.wings.android.navigationsampledsl.data.model.Memo
import com.websarva.wings.android.navigationsampledsl.databinding.FragmentMemoListBinding
import com.websarva.wings.android.navigationsampledsl.databinding.MemoListItemBinding
import com.websarva.wings.android.navigationsampledsl.navigation.MemoDetail
import com.websarva.wings.android.navigationsampledsl.ui.MemoListViewModel
import kotlinx.coroutines.flow.FlowCollector
import kotlinx.coroutines.launch

class MemoListFragment : Fragment() {
	private val _memoListViewModel: MemoListViewModel by viewModels()
	private var _fragmentMemoListBindingField: FragmentMemoListBinding? = null
	private val _fragmentMemoListBinding get() = _fragmentMemoListBindingField!!
	private lateinit var _adapter: MemoListAdapter


	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
		_fragmentMemoListBindingField = FragmentMemoListBinding.inflate(inflater)
		return _fragmentMemoListBinding.root
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		_adapter = MemoListAdapter()
		val layoutManager = LinearLayoutManager(requireActivity())
		_fragmentMemoListBinding.rvMenu.setLayoutManager(layoutManager)
		val dividerItemDecoration = DividerItemDecoration(requireActivity(), layoutManager.orientation)
		_fragmentMemoListBinding.rvMenu.addItemDecoration(dividerItemDecoration)
		_fragmentMemoListBinding.rvMenu.setAdapter(_adapter)

		val memoListFlow = _memoListViewModel.getMemoListFlow()
		viewLifecycleOwner.lifecycleScope.launch {
			repeatOnLifecycle(Lifecycle.State.STARTED) {
				memoListFlow.collect(MemoListCollector())
			}
		}
	}

	override fun onDestroyView() {
		super.onDestroyView()
		_fragmentMemoListBindingField = null
	}

	private inner class MemoListCollector : FlowCollector<List<Memo>> {
		override suspend fun emit(value: List<Memo>) {
			_adapter.submitList(value)
		}
	}

	private inner class MemoListAdapter : ListAdapter<Memo, MemoViewHolder?>(MemoDiff()) {
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MemoViewHolder {
			val inflater = LayoutInflater.from(requireActivity())
			val binding = MemoListItemBinding.inflate(inflater, parent, false)
			return MemoViewHolder(binding)
		}

		override fun onBindViewHolder(holder: MemoViewHolder, position: Int) {
			val memo: Memo = getItem(position)
			holder.bind(memo)
			holder.itemView.tag = memo.id
			holder.itemView.setOnClickListener(OnListItemClickListener())
		}
	}

	private inner class MemoDiff : DiffUtil.ItemCallback<Memo?>() {
		override fun areItemsTheSame(oldItem: Memo, newItem: Memo): Boolean {
			return oldItem.id == newItem.id
		}

		override fun areContentsTheSame(oldItem: Memo, newItem: Memo): Boolean {
			return oldItem == newItem
		}
	}

	private inner class MemoViewHolder(
		private val _binding: MemoListItemBinding
	) : RecyclerView.ViewHolder(_binding.root) {
		fun bind(memo: Memo) {
			_binding.memoTitleRow.text = memo.title
		}
	}

	private inner class OnListItemClickListener : View.OnClickListener {
		override fun onClick(view: View) {
			val memoId = view.tag as Long
			val navController = NavHostFragment.findNavController(this@MemoListFragment)
			navController.navigate(MemoDetail(memoId))
		}
	}
}
